### ISSUE TITLE
Suggesting new behaviour of weighted Fitness

### ISSUE THREAD

**Issue by profPlum** [DEAP/deap#377]
Although I can understand you reasoning for wanting to use lexical comparison, I believe that the resulting functionality is extremely misleading and that it could benefit from an overhaul.
This issue has been mentioned [before](https://github.com/DEAP/deap/issues/368) but I would like to bring it up again as I feel it is making many people implement algorithms they didn't intend to.

Surprising behaviour:
Specifically you ask for, and use weights, in your api/tutorial code despite them having no impact on the resulting algorithm aside from their sign. It seems to me you deap is essentially asking for boolean args that it takes in the form of floats. Additionally _weights_ have an implied continuous significance/nature, which gave me the impression that there must be a weighted sum going on behind the scenes.

Practical Implications:
Aside from the surprising nature of this behaviour, it also has practical implications that make secondary fitness scores more or less meaningless. In the vast majority of cases loss/fitness functions are going to be continuous, it could even be taken as a given. This means that the odds of two primary fitness scores ever being equal is effectively 0. Unless of course these fitnesses are rounded, but the need to do so is never implied or suggested and I don't believe it would occur to an average user.

**Comment by profPlum** [comment]
ALSO: I would like to add, that at the very least the Fitness class should throw an error when a weight is passed that isn't -1 or 1, that way no one will accidentally assume the weights are doing something that they aren't.

**Comment by rgerkin** [comment]
I agree completely.  I should also note that whether this is right way to do optimization is totally separate issue from whether this is the right way to implement and describe it.  It may very well be that we should only use signed booleans when there are multiple objectives.  But if so, calling them "weights" and letting them be floats has surely led many users including @profPlum and I on the wrong track.

**Comment by profPlum** [comment]
@rgerkin : That's fair, but they are both very important and related so I included them together. 

If we did want to keep using lexicographical comparison though: it should be indicated somewhere that the primary fitness should be rounded (likely to only a few points of precision). Otherwise you can assume they will never be equal.

**Comment by fmder** [comment]
In my opinion, lexicographical comparison is the most intuitive comparison for tuples, that how Python sorts them naturally. I think you have a fair point that it is not clear that DEAP considers fitnesses like tuples. Maybe instead of changing the behaviour of multi-objective fitnesses we should change the behaviour of single objective selection operators limiting them to 1 objective or something like that.

What would you expect?

**Comment by profPlum** [comment]
@fmder I can see why you'd prefer to use pure python as it does make for beautiful code; however I honestly find lexicographical comparison very strange. The vast majority of the time that I see/use sequence comparison I want/see numpy/R behaviour of element-wise comparison.

If you are really attached to lexicographical comparison though, I would at least _**strongly**_ suggest refraining from using 'weight' terminology. That terminology originates from 'weighted sums' and 'weighted averages' (which are really another type of sum); so to use it strongly implies some form of summation. I would suggest at the very least rename the 'weights' to 'coefficients' and throw errors when they are non-unit.

I believe that is most straightforward since while you are still using the 'weight' terminology I believe that my problem is already using 'single objective' selectors in the sense that they only are seeing the weighted sum. Because of that I honestly wasn't aware that there were 'single objective' and 'multi-objective' selectors until just now.

**Comment by profPlum** [comment]
tagging @joseortiz3 & @Nitrooo: how would you guys feel about that? Any second opinions?
